### DISCUSSION TITLE
Filter enabling syntactic sugar for using custom typst functions in .qmd file

### DISCUSSION THREAD

**Discussion by cl-roberts** [quarto-dev/quarto-cli#10381] [Show and Tell]
### Description

I made a [filter](https://github.com/cl-roberts/lua-filters/tree/main/typst-to-div) for using typst functions in a `.qmd` file by wrapping content in a div rather than disparate typst code blocks. This filter is heavily inspired by/generalized from [this one in the ams template](https://github.com/quarto-ext/typst-templates/blob/main/ams/_extensions/ams/ams.lua). 

### The usual way

I find it cumbersome to wrap content in typst code blocks, i.e.

````
```{=typst}
#block(
    fill:luma(230), 
    inset:16pt,
    outset:-8pt, 
    radius:4pt,
    [
```

...quarto content...

```{=typst}
    ]   
) 
```
````

### This way

Instead, I find it more natural to declare the function in a single typst block (or template file):

````
``` {=typst}
#let myblock(text) = block(text,
    fill:luma(230), 
    inset:16pt,
    outset:-8pt, 
    radius:4pt
  )
```
````

and call it using a div:

````
::: {.typst-to-div-myblock}
...quarto content...
:::
````

Note that the div class is named using the typst function name preceded by the project name (as determined as the parent directory to the `_extensions` folder) and an `'-'`.

**Comment by cscheid** [comment]
Cool! I'll note that Quarto [ships with some built-in functionality that's close to that](https://quarto.org/docs/output-formats/typst.html#typst-blocks):

````
---
format: typst
keep-typ: true
---


::: {.block fill="luma(100)"}

This has a luma of 100.

:::
````

<img width="221" alt="image" src="https://github.com/user-attachments/assets/d79d030b-26b3-401c-9cda-37983c50c034">

**Comment by jimjam-slam** [comment]
This is really cool, @cl-roberts — I was looking for exactly this functionality recently! I'm a bit too used to the HTML/CSS world where I define a style once 😅
